host/windows/syscall.c

```diff
@@ -480,142 +948,469 @@ ssize_t oe_syscall_writev_ocall(
+{
+    HANDLE hFind;
+    WIN32_FIND_DATAW FindFileData;
+    int dir_offs;
+    WCHAR* pdirpath;
+};
+
 uint64_t oe_syscall_opendir_ocall(const char* pathname)
 {
-    OE_UNUSED(pathname);
+    struct WIN_DIR_DATA* pdir =
+        (struct WIN_DIR_DATA*)calloc(1, sizeof(struct WIN_DIR_DATA));
+    WCHAR* wpathname = oe_syscall_path_to_win(pathname, "/*");
 
-    PANIC;
+    pdir->hFind = FindFirstFileW(wpathname, &pdir->FindFileData);
+    if (pdir->hFind == INVALID_HANDLE_VALUE)
+    {
+        free(wpathname);
+        free(pdir);
+        return 0;
+    }
+    pdir->dir_offs = 0;
+    pdir->pdirpath = wpathname;
+    return (uint64_t)pdir;
 }
 
 int oe_syscall_readdir_ocall(uint64_t dirp, struct oe_dirent* entry)
 {
-    OE_UNUSED(dirp);
-    OE_UNUSED(entry);
+    struct WIN_DIR_DATA* pdir = (struct WIN_DIR_DATA*)dirp;
+    int nlen = -1;
 
-    PANIC;
+    _set_errno(0);
+
+    if (!dirp || !entry)
+    {
+        _set_errno(OE_EINVAL);
+        return -1;
+    }
+
+    // Find file next doesn't return '.' because it shows up in opendir and we
+    // lose it but we know it is there, so we can just return it
+    if (pdir->dir_offs == 0)
+    {
+        entry->d_off = pdir->dir_offs++;
+        entry->d_type = OE_DT_DIR;
+        entry->d_reclen = sizeof(struct oe_dirent);
+        entry->d_name[0] = '.';
+        entry->d_name[1] = '\0';
+        return 0;
+    }
+
+    if (!FindNextFileW(pdir->hFind, &pdir->FindFileData))
+    {
+        DWORD winerr = GetLastError();
+
+        if (winerr == ERROR_NO_MORE_FILES)
+        {
+            /* Return 1 to indicate there no more entries. */
+            return 1;
+        }
+        else
+        {
+            _set_errno(_winerr_to_errno(winerr));
+            return -1;
+        }
+    }
+
+    nlen = WideCharToMultiByte(
+        CP_UTF8, 0, pdir->FindFileData.cFileName, -1, NULL, 0, NULL, NULL);
+    (void)WideCharToMultiByte(
+        CP_UTF8,
+        0,
+        pdir->FindFileData.cFileName,
+        nlen,
+        entry->d_name,
+        sizeof(entry->d_name),
+        NULL,
+        NULL);
+
+    entry->d_type = 0;
+    if (pdir->FindFileData.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)
+    {
+        entry->d_type = OE_DT_DIR;
+    }
+    else if (pdir->FindFileData.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT)
+    {
+        entry->d_type = OE_DT_LNK;
+    }
+    else
+    {
+        entry->d_type = OE_DT_REG;
+    }
+
+    entry->d_off = pdir->dir_offs++;
+    entry->d_reclen = sizeof(struct oe_dirent);
+
+    return 0;
 }
 
 void oe_syscall_rewinddir_ocall(uint64_t dirp)
 {
-    OE_UNUSED(dirp);
+    DWORD err = 0;
+    struct WIN_DIR_DATA* pdir = (struct WIN_DIR_DATA*)dirp;
+    WCHAR* wpathname = pdir->pdirpath;
+    // Undo abosolute path forcing again. We do this over because we need to
+    // preserve the allocation address for free.
+    if (wpathname[0] == '/' && wpathname[2] == ':')
+    {
+        wpathname++;
+    }
 
-    PANIC;
+    FindClose(pdir->hFind);
```

Comment:
Check return values here.
Also, use memset_s and memcpy_s instead of memset and memcpy on Windows